### PR TITLE
Replace Rc<RefCell<Request>> with Rc<Request>

### PR DESCRIPTION
**Pull request by nikkisquared**
To make using Request easier, I've changed it from an `Rc<RefCell<T>>` to just `Rc<T>`, and updated individual fields to `RefCell<T>` or `Cell<T>` as needed. Git seems to be telling me there's a conflict with request.rs, which I resolved locally, but I'm not sure what to do (if anything) to make it easier to merge.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9172)

### CODE REVIEW

File: components/net/fetch/request.rs [hunk 1]
```diff
@@ -233,17 +233,16 @@ fn fetch(request: Rc<RefCell<Request>>, cors_flag: bool) -> Response {
...
     Response::network_error()
 }
 
 /// [Basic fetch](https://fetch.spec.whatwg.org#basic-fetch)
-fn basic_fetch(request: Rc<RefCell<Request>>) -> Response {
+fn basic_fetch(request: Rc<Request>) -> Response {
 
-    let req = request.borrow();
-    let url = req.url_list.last().unwrap();
+    let url = request.current_url().clone();
```
Comment (KiChjang) [inline components/net/fetch/request.rs]: Why is there a clone here?

Comment (nikkisquared) [inline components/net/fetch/request.rs]: that was an oversight on my part, I've fixed it now!

Comment (KiChjang) [inline components/net/fetch/request.rs]: There are other places where `current_url()` is followed by a `clone()`. Make sure you fix those as well.

File: components/net/fetch/request.rs [hunk 2]
```diff
@@ -574,15 +573,19 @@ fn http_network_or_cache_fetch(request: Rc<RefCell<Request>>,
...
-        http_request.borrow_mut().headers.set(ContentLength(content_length_value));
+        http_request.headers.borrow_mut().set(ContentLength(content_length_value));
     }
 
     // Step 6
-    match http_request.borrow().referer {
+    let referer;
+    {
+        referer = http_request.referer.clone();
+    }
```
Comment (KiChjang) [inline components/net/fetch/request.rs]: Why is this here? Isn't it just enough to do `let referer = http_request.referer.clone()`?

Comment (nikkisquared) [inline components/net/fetch/request.rs]: That's weird, I swear I ran into this and fixed it. Must've been a similar legacy handling from elsewhere. I've updated it now- no clone is needed at all! Cloning was just to get around pointer lifetimes.

File: components/net/fetch/request.rs [hunk 3]
```diff
@@ -553,15 +552,15 @@ fn http_network_or_cache_fetch(request: Rc<RefCell<Request>>,
 
     // Step 1
     let http_request = if request_has_no_window &&
-        request.borrow().redirect_mode != RedirectMode::Follow {
+        request.redirect_mode != RedirectMode::Follow {
         request.clone()
     } else {
-        Rc::new(RefCell::new(request.borrow().clone()))
+        request.clone()
```
Comment (KiChjang) [inline components/net/fetch/request.rs]: Wait, sorry, this conditional check isn't even needed at all if both branches result in `request.clone()`.

Comment (KiChjang) [inline components/net/fetch/request.rs]: According to the spec, this really should be a copy of `request`. So `Rc::try_unwrap(request).ok().clone()` would be the way here.

Comment (nikkisquared) [inline components/net/fetch/request.rs]: guh, I thought I got that one before, too! I've reworked that whole bit, the spec requires the if statement which I added a todo about

File: components/net/fetch/request.rs [hunk 4]
```diff
@@ -544,24 +543,23 @@ fn http_fetch(request: Rc<RefCell<Request>>,
...
 
 /// [HTTP network or cache fetch](https://fetch.spec.whatwg.org#http-network-or-cache-fetch)
-fn http_network_or_cache_fetch(request: Rc<RefCell<Request>>,
+fn http_network_or_cache_fetch(request: Rc<Request>,
                                credentials_flag: bool,
                                authentication_fetch_flag: bool) -> Response {
 
+    // Step 1
+    // should this ever not be done?
+    let http_request = request.clone()
```
Comment (KiChjang) [inline components/net/fetch/request.rs]: After reading the spec, what you had before made sense, but not on the `else` branch. It should have been something like `Rc::try_unwrap(request).ok().clone()` instead.

**Comment by KiChjang** [comment]
@bors-servo delegate+
r=me post-nit

File: components/net/fetch/request.rs [hunk 5]
```diff
@@ -553,15 +552,19 @@ fn http_network_or_cache_fetch(request: Rc<RefCell<Request>>,
 
     // Step 1
     let http_request = if request_has_no_window &&
-        request.borrow().redirect_mode != RedirectMode::Follow {
+        request.redirect_mode != RedirectMode::Follow {
         request.clone()
     } else {
-        Rc::new(RefCell::new(request.borrow().clone()))
+        match Rc::try_unwrap(request.clone()) {
```
Comment (KiChjang) [inline components/net/fetch/request.rs]: I think this will fail every time you try to unwrap since there's at least 2 strong references to `request`.

Comment (KiChjang) [inline components/net/fetch/request.rs]: Also, a shorthand for this would be `Rc::new(Rc::try_unwrap(request).expect("Error msg here").clone());`

Comment (Ms2ger) [inline components/net/fetch/request.rs]: Rather than `try_unwrap`, you could probably use `Rc::new((*request).clone())`.

**Comment by jdm** [comment]
@bors-servo: r+
Solid.

**Comment by KiChjang** [comment]
@bors-servo retry
* #9124, #9033, #9163
